### DISCUSSION TITLE
CloudWatch Label Improvements

### DISCUSSION THREAD

**Discussion by jpmatusik** [grafana/grafana#43386] [Dashboards, Panels & Visualizations]
## Issue
I am trying to use the CloudWatch data source and am running into 2 issues. Both can be solved with the same solutions (i hope).

I ran into the same issue as described in #26794 

Which is basically when:
1. you run a query with a wildcard search. for example a wildcard search on product_id...
```
REMOVE_EMPTY(SEARCH('{"namespace","app","env","product_id","user_id"} MetricName="name" "app"="api" "env"="test" "user_id"="the-user-id"', 'Average', 60))
```
2. only a single metric is returned

If 1 and 2 are satisfied then the label for the returned series will be the metric name.

The other issue I run into is when:
1. I run a query with multiple wildcards wildcard search. for example a wildcard search on product_id and user_id...
```
REMOVE_EMPTY(SEARCH('{"namespace","app","env","product_id","user_id"} MetricName="name" "app"="api" "env"="test"', 'Average', 60))
```
2. multiple series are returned
3. I run the `Labels to fields` transformation

When I do this, what happens is the columns that get created are hard to work with. As expected columns are created for each dimension used. As expected the values for the columns for the dimensions using the exact match are populated with the value that was searched for. Where things get weird are the columns that get created for the dimensions that were used in wildcard matches.

What happens all the columns for the wildcard dimensions are all the same. The values for the wildcard dimensions get concatenated and replicated across all the columns.

So in the above example with the two wildcard dimensions being `product_id` and `user_id` you see something like this.

user_id                | product_id
---------------- | ----------------
user-a product a | user-a product a
user-a product b | user-a product b
user-a product c | user-a product c
user-b product z | user-b product z
user-b product y | user-b product y
user-c product g | user-c product g

## Question
Can this be improved? Specifically if a wildcard search returns only 1 series, can the label include the value of the dimension that was found? And in the case of multiple wildcard dimension matches -> `Labels to fields`, can the columns not concatenate all values?

## Motivation

I'm looking at the API for GetMetricData and I see that dynamic labels can be added to a query.

https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/graph-dynamic-labels.html

Of specific interest is ${PROP('Dim.dimension_name')}.

In boto3, using it looks like this with product_id as a wildcard.

```
import boto3
from datetime import datetime

client = boto3.client("cloudwatch", region_name='us-east-2')

response = client.get_metric_data(
    MetricDataQueries=[
        {
            'Id': 'a',
            'Expression': '''REMOVE_EMPTY(SEARCH('{"namespace","app","env","product_id","user_id"} MetricName="name" "app"="api" "env"="test" "user_id"="the-user-id"', 'Average', 60))''',
            'Label': "product_id=${PROP('Dim.product_id')}\nuser_id=${PROP('Dim.user_id')}\nenv=${PROP('Dim.env')}",
        },
    ],

    StartTime=datetime(2021, 12, 21, 0),
    EndTime=datetime(2021, 12, 22),
)

print(len(response['MetricDataResults']))
print(response['MetricDataResults'][0]['Label'])
```
where in this case the printed output look like this 
```
# from the len
1 
# from the label
product_id=the-product-id
user_id=the-user-id
env=test
```

Seems to me that manually constructing the label could be useful here.

**Comment by fridgepoet** [comment]
Thanks! We aim to get this into Grafana 9.0 https://github.com/grafana/grafana/issues/48434

**Comment by sarahzinger** [comment]
I think this has been addressed with the release of dynamic labels but if we missed anything feel free to let us know!
